### PR TITLE
DO NOT MERGE: Split the Camera and the Avatar and introduce the UserInputMapper

### PR DESCRIPTION
**Pull request by samcake**
This is a work in progress! but we need to test this, prototype more and do it again...

- Split the Camera from the Avatar, meaning:
    - The camera 1st and 3rd person modes are merged into the "Camera_mode_avatar", the only difference between the 2 modes is the boom distance of the Camera.
   - The Camera position is starting from the same Avatar hips position. That's the root frame of the Camera. THis is where the yaw of the Camera is injected
   - The pivot of the Camera is the frame placed at the default eye position of the avatar. This is where the pitch of the camera is injected
 - Lastly, the View Frame is moved along the Z axis of the boom distance

- The userInputMapper is the class doing the mapping between the user input devices among others the default keyboardMouseDevice and the "logical" channel of the application. this is where all the different places that can "control" the application interface together in a single set of channels that the whole application (and js) can pull from.

-  The bridge between the userInputMapper and the Camera and MyAVatar is happening in the LookMoveController. For now it;s a hard coded behavior but the plan is a js function could take over that task so user can transform the way the Camera and the avatar are moving from the user inputs.

So please test and prototype and let me know how it works  or not for you.

Wednesday 4/29: it s somewhat good to go right now
Still issues with the mac book touch pad
And the rear view mirror camera

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: a1e19cb88d3e0feeb25e25d3676bf32d13020f05 available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -2251,13 +2147,15 @@ void Application::cameraMenuChanged() {
             _myCamera.setMode(CAMERA_MODE_MIRROR);
         }
     } else if (Menu::getInstance()->isOptionChecked(MenuOption::FirstPerson)) {
-        if (_myCamera.getMode() != CAMERA_MODE_FIRST_PERSON) {
+        //if (_myCamera.getMode() != CAMERA_MODE_FIRST_PERSON) {
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: do you want to remove this? if so, let's remove it.

File: interface/src/Camera.cpp [hunk 2]
```diff
@@ -178,3 +198,160 @@ void Camera::keepLookingAt(const glm::vec3& point) {
...
+    _transform.postRotate(_pivotOrientation);
+
+    _transform.postTranslate(_viewPosition);
+    _transform.postRotate(_viewOrientation);
+
+    _transformInvalid = false;
+}
+
+
+void LookMoveController::update(float deltaTime, const UserInputMapper& mapper, MyAvatar& avatar, Camera& camera) {
```
Comment (ZappoMan) [inline interface/src/Camera.cpp]: Coding standard: move to dedicated file

File: interface/src/Camera.h [hunk 3]
```diff
@@ -37,11 +42,16 @@ class Camera : public QObject {
     Q_PROPERTY(glm::quat orientation READ getOrientation WRITE setOrientation)
     Q_PROPERTY(QString mode READ getModeString WRITE setModeString)
 public:
+    typedef glm::vec3 Vec3;
```
Comment (ZappoMan) [inline interface/src/Camera.h]: We really just just be using the glm::vec3 types for consistency.

File: interface/src/ui/UserInputMapper.h [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+        void touchUpdateEvent(const TouchEvent& event);
+
+        void wheelEvent(QWheelEvent* event);
+
+        ButtonPressedMap _buttonPressedMap;
+        mutable AxisStateMap _axisStateMap;
+
+        QPoint _lastCursor;
+
+        float getButton(int channel) const {
```
Comment (ZappoMan) [inline interface/src/ui/UserInputMapper.h]: move to cpp

File: interface/src/ui/UserInputMapper.h [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+            if (!_buttonPressedMap.empty()) {
+                if (_buttonPressedMap.find(channel) != _buttonPressedMap.end()) {
+                    return 1.0f;
+                } else {
+                    return 0.0f;
+                }
+            }
+            return 0.0f;
+        }
+        float getAxis(int channel) const {
```
Comment (ZappoMan) [inline interface/src/ui/UserInputMapper.h]: move to cpp

File: interface/src/ui/UserInputMapper.h [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+
+    
+        // Reserverd channel for the standard keyboard and mouse
+        const static uint16 DEFAULT_DESKTOP_DEVICE = 1;
+
+        // Let's make it easy for Qt because we assume we love Qt forever
+        static InputID makeInputID(Qt::Key code) {
+            return InputID(DEFAULT_DESKTOP_DEVICE, code & KEYBOARD_MASK, ChannelType::BUTTON);
+        }
+        static InputID makeInputID(Qt::MouseButton code) {
```
Comment (ZappoMan) [inline interface/src/ui/UserInputMapper.h]: move to cpp
coding standard formating

File: interface/src/ui/UserInputMapper.h [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+        }
+
+
+        KeyboardMouseDevice() {}
+
+        bool buttonGetter(int timestamp, const InputID& input) {
+            return getButton(input._channel);
+        }
+
+        void registerInputs(UserInputMapper& mapper) {
```
Comment (ZappoMan) [inline interface/src/ui/UserInputMapper.h]: move to cpp
coding standard formating

File: interface/src/ui/UserInputMapper.h [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+            mapper.registerInputAxis(makeInputID(MOUSE_AXIS_Y_NEG), [this] (const InputID& input, int timestamp) -> float { return this->getAxis(input._channel); });
+            mapper.registerInputAxis(makeInputID(MOUSE_AXIS_Y_POS), [this] (const InputID& input, int timestamp) -> float { return this->getAxis(input._channel); });
+            mapper.registerInputAxis(makeInputID(MOUSE_AXIS_WHEEL_X_NEG), [this] (const InputID& input, int timestamp) -> float { return this->getAxis(input._channel); });
+            mapper.registerInputAxis(makeInputID(MOUSE_AXIS_WHEEL_X_POS), [this] (const InputID& input, int timestamp) -> float { return this->getAxis(input._channel); });
+            mapper.registerInputAxis(makeInputID(MOUSE_AXIS_WHEEL_Y_NEG), [this] (const InputID& input, int timestamp) -> float { return this->getAxis(input._channel); });
+            mapper.registerInputAxis(makeInputID(MOUSE_AXIS_WHEEL_Y_POS), [this] (const InputID& input, int timestamp) -> float { return this->getAxis(input._channel); });
+
+        }
+
+        void assignDefaultInputMapping(UserInputMapper& mapper) {
```
Comment (ZappoMan) [inline interface/src/ui/UserInputMapper.h]: ditto

File: interface/src/ui/UserInputMapper.h [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+    InputToMoModifiersMap _inputToModifiersMap;
+
+    class InputChannelMapping {
+    public:
+        InputChannel _input;
+        Channel _output = LONGITUDINAL_BACKWARD;
+        float _scale = 0.0f;
+
+        InputChannelMapping() {}
+        InputChannelMapping(const InputChannel& input, Channel output, float scale = 1.0f) :
```
Comment (ZappoMan) [inline interface/src/ui/UserInputMapper.h]: move to cpp

File: interface/src/ui/UserInputMapper.h [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+        _keyboardMouseDevice.registerInputs(*this);
+        assignDefaultInputChannelMap();
+        assignDefaulChannelUnitScales(); }
+
+    // This call reset the Controller to the default key map as defined below:
+    void assignDefaultInputChannelMap() {
+        _keyboardMouseDevice.assignDefaultInputMapping((*this));
+    }
+
+    void assignDefaulChannelUnitScales() {
```
Comment (ZappoMan) [inline interface/src/ui/UserInputMapper.h]: move to cpp

File: interface/src/ui/UserInputMapper.h [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+        NUM_CHANNELS,
+    };
+
+    float getOuput(Channel channel) const { return _channelStates[channel]; }
+
+
+    void addInputChannel(Channel channel, const InputChannel& input, float scale = 1.0f);
+
+
+    UserInputMapper() {
```
Comment (ZappoMan) [inline interface/src/ui/UserInputMapper.h]: move to cpp

File: interface/src/ui/UserInputMapper.h [hunk 12]
```diff
@@ -0,0 +1,381 @@
...
+        InputID _id;
+        Modifiers _modifiers;
+
+        explicit InputChannel() {}
+        explicit InputChannel(const InputID& id) : _id(id) {}
+        explicit InputChannel(const InputID& id, const InputID& mod) : _id(id) { addModifier(mod); }
+        explicit InputChannel(const InputChannel& src) : _id(src._id), _modifiers(src._modifiers) {}
+        InputChannel& operator = (const InputChannel& src) { _id = src._id; _modifiers = src._modifiers; return (*this); }
+
+        void addModifier(const InputID& modifier) {
```
Comment (ZappoMan) [inline interface/src/ui/UserInputMapper.h]: move to cpp

**Comment by hifi-gustavo** [comment]
Build console output for commit: 679279a04d7ca22dbe4a92cd3185c93f0d7fa824 available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 067ba19bb2f0ae9b1fd350b8ba70577fac2d4eb2 available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b97388120a63d65ec4e25baa1ae81e0e9511e619 available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1be670d45d74afc2c1d65391fb41937f16dd8b1a available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b1a017e5489588427caa1534ceb40f5e98211814 available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3fce232510a1c096cf3cb7a39a39073010ff4b97 available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d22d92d43305b624edd88640be014333725e55fc available at: https://gist.github.com/d51dd177bbd65b4dd00c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bf5e2c46a9decd0434935b8c3993cab19e8c3bba available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

File: interface/src/ui/LookMoveController.cpp [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    if (pitch != 0) {
+        camera.addPitch(pitch);
+    }
+
+    float boom = mapper.getOutput(UserInputMapper::BOOM_OUT) - mapper.getOutput(UserInputMapper::BOOM_IN);
+    if (boom != 0) {
+        camera.addBoom(boom);
+    }
+
+    glm::quat q = glm::slerp(camera.getRootOrientation(), avatar.getOrientation(), 0.9f);
```
Comment (ZappoMan) [inline interface/src/ui/LookMoveController.cpp]: coding standard: variable name

File: interface/src/ui/LookMoveController.cpp [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    avatar.setOrientation(q);
+    glm::quat qh = glm::slerp(camera.getRootOrientation() * camera.getPivotOrientation(), avatar.getHeadOrientation(), 0.5f);
```
Comment (ZappoMan) [inline interface/src/ui/LookMoveController.cpp]: coding standard: variable name

**Comment by hifi-gustavo** [comment]
Build console output for commit: 86e1511e28fb62587402298a499acb46ccb20570 available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 37cdcd9b9d228d55cf9df52cf2ebc77b6cf5ff5f available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 040b91eca226fc8a5efe024b5dffbb266b014a4e available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by samcake** [comment]
Please test, review and maybe merge...
Thanks

**Comment by hifi-gustavo** [comment]
Build console output for commit: 97bc3326f90b5a28d7f4eb6c371da40b5941eaa4 available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

File: interface/src/Camera.cpp [hunk 15]
```diff
@@ -177,3 +197,126 @@ void Camera::keepLookingAt(const glm::vec3& point) {
...
+}
+void Camera::setBoom(float distance) {
+    _boom = clampRange(distance, 0.0f, MAX_BOOM);
+    invalidRootPivotView();
+}
+float Camera::getBoom() const {
+    return _boom;
+}
+
+void Camera::setRootPosition(const Vec3& position) {
```
Comment (ZappoMan) [inline interface/src/Camera.cpp]: I really think we should stick with the glm::vec3 types here... we use the fully qualified glm types throughout the code base... why you got to be different!?

File: interface/src/Camera.cpp [hunk 16]
```diff
@@ -177,3 +197,126 @@ void Camera::keepLookingAt(const glm::vec3& point) {
...
+    _rootPosition = position;
+    _transformInvalid = true;
+}
+void Camera::setPivotPosition(const Vec3& position) {
```
Comment (ZappoMan) [inline interface/src/Camera.cpp]: glm::vec3

File: interface/src/Camera.cpp [hunk 17]
```diff
@@ -177,3 +197,126 @@ void Camera::keepLookingAt(const glm::vec3& point) {
...
+    _pivotPosition = position;
+    _transformInvalid = true;
+}
+void Camera::setViewPosition(const Vec3& position) {
```
Comment (ZappoMan) [inline interface/src/Camera.cpp]: glm::vec3

File: interface/src/Camera.cpp [hunk 18]
```diff
@@ -177,3 +197,126 @@ void Camera::keepLookingAt(const glm::vec3& point) {
...
+}
+void Camera::setRootOrientation(const Quat& position) {
+    _rootOrientation = position;
+    _transformInvalid = true;
+}
+void Camera::setPivotOrientation(const Quat& position) {
+    _pivotOrientation = position;
+    _transformInvalid = true;
+}
+void Camera::setViewOrientation(const Quat& position) {
```
Comment (ZappoMan) [inline interface/src/Camera.cpp]: glm::quat

File: interface/src/Camera.cpp [hunk 19]
```diff
@@ -177,3 +197,126 @@ void Camera::keepLookingAt(const glm::vec3& point) {
...
+}
+void Camera::setViewPosition(const Vec3& position) {
+    _viewPosition = position;
+    _transformInvalid = true;
+}
+void Camera::setRootOrientation(const Quat& position) {
+    _rootOrientation = position;
+    _transformInvalid = true;
+}
+void Camera::setPivotOrientation(const Quat& position) {
```
Comment (ZappoMan) [inline interface/src/Camera.cpp]: glm::quat

File: interface/src/Camera.cpp [hunk 20]
```diff
@@ -177,3 +197,126 @@ void Camera::keepLookingAt(const glm::vec3& point) {
...
+}
+void Camera::setPivotPosition(const Vec3& position) {
+    _pivotPosition = position;
+    _transformInvalid = true;
+}
+void Camera::setViewPosition(const Vec3& position) {
+    _viewPosition = position;
+    _transformInvalid = true;
+}
+void Camera::setRootOrientation(const Quat& position) {
```
Comment (ZappoMan) [inline interface/src/Camera.cpp]: glm::quat

File: interface/src/Camera.h [hunk 21]
```diff
@@ -37,11 +43,16 @@ class Camera : public QObject {
...
+    typedef glm::quat Quat;
```
Comment (ZappoMan) [inline interface/src/Camera.h]: I vote we get rid of this and use the glm type

File: interface/src/Camera.h [hunk 22]
```diff
@@ -90,6 +114,43 @@ public slots:
...
+    float getSnapToPivotBoomDistance() const { return _snapToPivotBoomDistance; }
+
+    const glm::vec3& getRootPosition() const { if (_rootPivotViewInvalid) { updateRootPivotView(); } return _rootPosition; }
+    const glm::vec3& getPivotPosition() const { if (_rootPivotViewInvalid) { updateRootPivotView(); } return _pivotPosition; }
+    const glm::vec3& getViewPosition() const { if (_rootPivotViewInvalid) { updateRootPivotView(); } return _viewPosition; }
+    const Quat& getRootOrientation() const { if (_rootPivotViewInvalid) { updateRootPivotView(); } return _rootOrientation; }
+    const Quat& getPivotOrientation() const { if (_rootPivotViewInvalid) { updateRootPivotView(); } return _pivotOrientation; }
+    const Quat& getViewOrientation() const { if (_rootPivotViewInvalid) { updateRootPivotView(); } return _viewOrientation; }
+
+    void setRootPosition(const Vec3& position);
```
Comment (ZappoMan) [inline interface/src/Camera.h]: glm::vec3

File: interface/src/Camera.h [hunk 23]
```diff
@@ -90,6 +114,43 @@ public slots:
...
+    void setPivotPosition(const Vec3& position);
+    void setViewPosition(const Vec3& position);
+    void setRootOrientation(const Quat& orientation);
```
Comment (ZappoMan) [inline interface/src/Camera.h]: glm::quat

File: interface/src/Camera.h [hunk 24]
```diff
@@ -90,6 +114,43 @@ public slots:
...
+    void setPivotOrientation(const Quat& orientation);
```
Comment (ZappoMan) [inline interface/src/Camera.h]: glm::quat

File: interface/src/Camera.h [hunk 25]
```diff
@@ -90,6 +114,43 @@ public slots:
...
+    void setViewOrientation(const Quat& orientation);
```
Comment (ZappoMan) [inline interface/src/Camera.h]: glm::quat

File: interface/src/Camera.h [hunk 26]
```diff
@@ -90,6 +114,43 @@ public slots:
...
+
+    const glm::vec3& getRootPosition() const { if (_rootPivotViewInvalid) { updateRootPivotView(); } return _rootPosition; }
+    const glm::vec3& getPivotPosition() const { if (_rootPivotViewInvalid) { updateRootPivotView(); } return _pivotPosition; }
+    const glm::vec3& getViewPosition() const { if (_rootPivotViewInvalid) { updateRootPivotView(); } return _viewPosition; }
+    const Quat& getRootOrientation() const { if (_rootPivotViewInvalid) { updateRootPivotView(); } return _rootOrientation; }
+    const Quat& getPivotOrientation() const { if (_rootPivotViewInvalid) { updateRootPivotView(); } return _pivotOrientation; }
+    const Quat& getViewOrientation() const { if (_rootPivotViewInvalid) { updateRootPivotView(); } return _viewOrientation; }
+
+    void setRootPosition(const Vec3& position);
+    void setPivotPosition(const Vec3& position);
```
Comment (ZappoMan) [inline interface/src/Camera.h]: glm::vec3

File: interface/src/Camera.h [hunk 27]
```diff
@@ -90,6 +114,43 @@ public slots:
...
+    void setViewPosition(const Vec3& position);
```
Comment (ZappoMan) [inline interface/src/Camera.h]: glm::vec3

File: interface/src/Camera.h [hunk 28]
```diff
@@ -107,6 +168,34 @@ public slots:
     glm::quat _hmdRotation;
     bool _isKeepLookingAt;
     glm::vec3 _lookingAt;
+
+    // Controlling parameters
+    Vec3 _translate{ 0.0f };
```
Comment (ZappoMan) [inline interface/src/Camera.h]: glm::vec3

File: interface/src/Camera.h [hunk 29]
```diff
@@ -107,6 +168,34 @@ public slots:
...
+    float _yaw{ 0.0f };
+    float _pitch{ 0.0f };
+    float _boom{ 0.0f };
+
+    float _snapToPivotBoomDistance{ DEFAULT_SNAP_TO_PIVOT_BOOM_DISTANCE }; // Distance before snapping the boom to pivot
+
+    // Root / Pivot / View is the transform chain defining the full camera transform
+    mutable Vec3 _rootPosition;
```
Comment (ZappoMan) [inline interface/src/Camera.h]: glm::vec3

File: interface/src/ui/UserInputMapper.cpp [hunk 30]
```diff
@@ -0,0 +1,340 @@
...
+        _lastCursor = currentPos;
+    }
+}
+
+UserInputMapper::Input KeyboardMouseDevice::makeInput(Qt::Key code) {
+    return UserInputMapper::Input(DEFAULT_DESKTOP_DEVICE, code & KEYBOARD_MASK, UserInputMapper::ChannelType::BUTTON);
+}
+UserInputMapper::Input KeyboardMouseDevice::makeInput(Qt::MouseButton code) {
+    switch (code) {
+        case Qt::LeftButton: return UserInputMapper::Input(DEFAULT_DESKTOP_DEVICE, MOUSE_BUTTON_LEFT, UserInputMapper::ChannelType::BUTTON);
```
Comment (ZappoMan) [inline interface/src/ui/UserInputMapper.cpp]: coding standard: switch statement formatting see https://docs.highfidelity.com/v1.0/docs/coding-standard#4-1-9-switch-case-statements-

File: interface/src/ui/UserInputMapper.cpp [hunk 31]
```diff
@@ -0,0 +1,340 @@
...
+        }
+
+        // If any modified version is on then dig deeper:
+        if (!possibilities.empty()) {
+            // definitely not the raw input which is active since there is modifiers
+            if (!inputMapping.hasModifiers()) {
+               enabled = false;
+            } else {
+                // We assume that the first possible is the one that won
+            //    if (inputMapping._input._modifiers != possibilities[0]) {
```
Comment (ZappoMan) [inline interface/src/ui/UserInputMapper.cpp]: indent? also maybe remove this dead code and/or add TODO note

File: libraries/gpu/src/gpu/GLBackendOutput.cpp [hunk 32]
```diff
@@ -38,7 +38,8 @@ GLBackend::GLFramebuffer* GLBackend::syncGPUObject(const Framebuffer& framebuffe
         (void) CHECK_GL_ERROR();
 
         glBindFramebuffer(GL_FRAMEBUFFER, fbo);
-
+        CHECK_GL_ERROR();
```
Comment (ZappoMan) [inline libraries/gpu/src/gpu/GLBackendOutput.cpp]: won't this create warnings unless we add the (void)

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3a6ba3644251ddc5dc5dd122ed1c0f3759139236 available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 82b66811e68d689e1b0705917e7256e259e3930f available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by ZappoMan** [comment]
@samcake - need to pull from upstream

File: interface/src/Camera.cpp [hunk 33]
```diff
@@ -99,6 +107,16 @@ float Camera::getFarClip() const {
...
+    if  (mode == CAMERA_MODE_FIRST_PERSON) {
+        mode = CAMERA_MODE_AVATAR;
+        setBoom(0.f); 
+    } else if (mode == CAMERA_MODE_THIRD_PERSON) {
+        mode = CAMERA_MODE_AVATAR;
+        setBoom(DEFAULT_3RD_PERSON_BOOM_DISTANCE); 
+    }
+    if (mode == CAMERA_MODE_AVATAR) {
+        setNearClip(_snapToPivotBoomDistance * 0.5f);
+    }
```
Comment (AndrewMeadows) [inline interface/src/Camera.cpp]: 0.f --> 0.0f on line 112.

File: interface/src/ui/UserInputMapper.cpp [hunk 34]
```diff
@@ -0,0 +1,335 @@
...
+    return 0.0f;
+}
+float KeyboardMouseDevice::getAxis(int channel) const {
+    auto axis = _axisStateMap.find(channel);
+    if (axis != _axisStateMap.end()) {
+        return (*axis).second;
+    } else {
+        return 0.f;
+    }
+}
```
Comment (AndrewMeadows) [inline interface/src/ui/UserInputMapper.cpp]: 0.f --> 0.0f at line 199

**Comment by hifi-gustavo** [comment]
Build console output for commit: 18178fd2005c456ce981bb8a9e766b2cbebc378d available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ceccfce3e59df7fa376bd2fe2722aa65d00ba018 available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9a306f301735d8966511e9e2a3f99eb1b5e4402d available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: df137c1b3d3f3fe11332866e4b4b01245557bc39 available at: https://gist.github.com/76d9bc66a07f18eb2c4b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4711.dmg

**Comment by samcake** [comment]
DO not merge, the HMD experience is not working with the new camera design...

**Comment by ZappoMan** [comment]
@samcake - maybe close this PR for now?

**Pull request closed without merging**